### PR TITLE
fix: commas in column name now work  in sort key

### LINKED ISSUE #4762: Panic: "Primary key column(s) found that don't appear in the catalog sort key" found in iox-shared-compactor-3-0
Only happen in `iox-shared-compactor-3-0`

```
level=error msg="Thread panic" panic_info="panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [[\"tatraffic_vnstat_live_30s_rx,host\", \"taupgradesavail_apt,host\"]]', schema/src/sort.rs:217:13" target="panic_logging" location="panic_logging/src/lib.rs:107" time=1654095960183115430
thread 'tokio-runtime-worker' panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [["tatraffic_vnstat_live_30s_rx,host", "taupgradesavail_apt,host"]]', schema/src/sort.rs:217:13

level=error msg="Thread panic" panic_info="panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [[\"tatraffic_vnstat_live_30s_rx,host\", \"taupgradesavail_apt,host\"]]', schema/src/sort.rs:217:13" target="panic_logging" location="panic_logging/src/lib.rs:107" time=1654095960185830698
thread 'tokio-runtime-worker' panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [["tatraffic_vnstat_live_30s_rx,host", "taupgradesavail_apt,host"]]', schema/src/sort.rs:217:13

level=error msg="Thread panic" panic_info="panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [[\"tartraffic_vnstat_live_30s_rx,host\", \"tarupgradesavail_apt,host\", \"tatraffic_vnstat_live_30s_rx,host\", \"taupgradesavail_apt,host\"]]', schema/src/sort.rs:217:13" target="panic_logging" location="panic_logging/src/lib.rs:107" time=1654095960525857724
thread 'tokio-runtime-worker' panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [["tartraffic_vnstat_live_30s_rx,host", "tarupgradesavail_apt,host", "tatraffic_vnstat_live_30s_rx,host", "taupgradesavail_apt,host"]]', schema/src/sort.rs:217:13

level=error msg="Thread panic" panic_info="panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [[\"tatraffic_vnstat_live_30s_rx,host\", \"taupgradesavail_apt,host\", \"tawan_rx_bytes,host\", \"tawan_tx_bytes,host\"]]', schema/src/sort.rs:217:13" target="panic_logging" location="panic_logging/src/lib.rs:107" time=1654095960531400333
thread 'tokio-runtime-worker' panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [["tatraffic_vnstat_live_30s_rx,host", "taupgradesavail_apt,host", "tawan_rx_bytes,host", "tawan_tx_bytes,host"]]', schema/src/sort.rs:217:13

level=error msg="Thread panic" panic_info="panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [[\"tartraffic_vnstat_live_30s_rx,host\", \"tarupgradesavail_apt,host\"]]', schema/src/sort.rs:217:13" target="panic_logging" location="panic_logging/src/lib.rs:107" time=1654095960535641050
thread 'tokio-runtime-worker' panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [["tartraffic_vnstat_live_30s_rx,host", "tarupgradesavail_apt,host"]]', schema/src/sort.rs:217:13

level=error msg="Thread panic" panic_info="panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [[\"tardocker_netstat_combined,target\", \"tartraffic_vnstat_live_30s_rx,host\", \"tarupgradesavail_apt,host\"]]', schema/src/sort.rs:217:13" target="panic_logging" location="panic_logging/src/lib.rs:107" time=1654095960560062547
thread 'tokio-runtime-worker' panicked at 'Primary key column(s) found that don't appear in the catalog sort key: [["tardocker_netstat_combined,target", "tartraffic_vnstat_live_30s_rx,host", "tarupgradesavail_apt,host"]]', schema/src/sort.rs:217:13



```

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #4762

The fix is to add escapes in front of commas at serializing the sort keys (aka WRITE) and remove the escapes at deserializing step (aka READ). The PR has many new small functions and tests  but the main changes to invoke them is at READ and WRITE that I have added comments for reviewers to follow easier.

Since we want to add and remove escapes for commas specifically, I write our own code instead of use csv crate. Let me know what you think @alamb 

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: data_types/src/lib.rs [hunk 1]
```diff
@@ -674,15 +674,23 @@ pub struct Partition {
     pub partition_key: String,
     /// The sort key for the partition. Should be computed on the first persist operation for
     /// this partition and updated if new tag columns are added.
+    /// This is a string of concactenated column names of the sort key, each separated by a comma. If
+    /// there are commas in a column name, they must have escape in front.
+    /// Example:
+    ///    Three columns in the sort key named: "col1", "col2_with,comma", "time"
+    ///    -> the sort_key is Some("col1,col2_with\\,comma,time")
```
Comment (NGA-TRAN) [inline data_types/src/lib.rs:681]:
```suggestion
    ///    ->  Value of the sort_key is Some("col1,col2_with\\,comma,time")
```

File: data_types/src/lib.rs [hunk 2]
```diff
@@ -3131,4 +3139,21 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_partition() {
+        let partition = Partition {
+            id: PartitionId::new(1),
+            sequencer_id: SequencerId::new(1),
+            table_id: TableId::new(1),
+            partition_key: "whatever".to_string(),
+            sort_key: Some("a\\,b,c,d\\,".to_string()),
```
Comment (NGA-TRAN) [inline data_types/src/lib.rs:3150]:
```suggestion
            sort_key: Some("a\\,b,c,d\\,".to_string()), // A string with escapes
```

File: data_types/src/lib.rs [hunk 3]
```diff
@@ -3131,4 +3139,21 @@ mod tests {
...
+        };
+
+        assert_eq!(partition.sort_key, Some("a\\,b,c,d\\,".to_string()));
```
Comment (NGA-TRAN) [inline data_types/src/lib.rs:3153]:
```suggestion
      // Same string with escapes
        assert_eq!(partition.sort_key, Some("a\\,b,c,d\\,".to_string()));
```

File: ingester/src/data.rs [hunk 4]
```diff
@@ -327,7 +327,7 @@ impl Persister for IngesterData {
 
             // Update the sort key in the catalog if there are additional columns
             if let Some(new_sort_key) = sort_key_update {
-                let sort_key_string = new_sort_key.to_columns();
+                let sort_key_string = new_sort_key.to_columns_string();
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:330]: Rename this to make it clear it is a string of columns

File: query_functions/src/regex.rs [hunk 5]
```diff
@@ -152,7 +152,7 @@ fn is_valid_character_after_escape(c: char) -> bool {
 /// golang, used by the influx storage rpc.
 ///
 /// See <https://github.com/rust-lang/regex/issues/501> for more details
-fn clean_non_meta_escapes(pattern: &str) -> String {
+pub fn clean_non_meta_escapes(pattern: &str) -> String {
```
Comment (NGA-TRAN) [inline query_functions/src/regex.rs:155]:
```suggestion
fn clean_non_meta_escapes(pattern: &str) -> String {
```

File: schema/Cargo.toml [hunk 6]
```diff
@@ -12,3 +12,4 @@ indexmap = { version = "1.8", features = ["std"] }
 itertools = "0.10.1"
 snafu = "0.7"
 workspace-hack = { path = "../workspace-hack"}
+
```
Comment (NGA-TRAN) [inline schema/Cargo.toml:15]:
```suggestion
```

File: schema/src/sort.rs [hunk 7]
```diff
@@ -137,6 +137,27 @@ impl SortKey {
         }
     }
 
+    pub fn from_columns_string_with_escape(cols_string: &str) -> Self {
```
Comment (NGA-TRAN) [inline schema/src/sort.rs:140]:
```suggestion
    /// Create a sort key from a string of columns with escapes
    pub fn from_columns_string_with_escape(cols_string: &str) -> Self {
```

File: schema/src/sort.rs [hunk 8]
```diff
@@ -945,4 +1066,78 @@ mod tests {
         let key_2 = SortKey::from_columns(vec!["a", TIME_COLUMN_NAME]);
         assert!(key_1.size() < key_2.size());
     }
+
+    #[test]
+    fn test_escape() {
```
Comment (alamb) [inline schema/src/sort.rs:1071]: What about partition keys that contain `\`? Can you please add a test for them as well?

Like `Over\\Under`?

**Review by alamb — APPROVED** [review]
Approving this PR as it does appear to fix the bug, as demonstrated by tests, but I definitely have concerns about the implementation

# Questions
1. What is the story for the data already in the production database?

# Concerns
I am worried about 
1. the custom escaping and parsing may be brittle / have edge cases(for example, does the code handle escaping escapes) 
2. There is a lot of custom parsing / serialization code


> Since we want to add and remove escapes for commas specifically, I write our own code instead of use csv crate. Let me know what you think @alamb

I don't really understand this choice. I think this PR would contain less code and be less likely to have corner cases if it used the `csv` crate.

Here is an example I whipped up on the rust playground:  https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f6fdfc5cf0a65b408150a0b9bdca463

It produces output like this:
```
serialized_data is: col1,"col2,with"

deserialized_column_names is: ["col1", "col2,with"]
```

Specifically, it serializes data using csv style quoting and seems to require much less code

File: schema/src/sort.rs (hunk 8 above)
Comment (NGA-TRAN) [reply schema/src/sort.rs:1071]: I have added 2 tests below (commit shortly) . The idea is,  as in its name `add_escape_before_comma`, we do not touch other escape or characters, just the commas 

```
let result = SortKey::add_escape_before_comma("\\?");
assert_eq!(result, "\\?");

let result = SortKey::add_escape_before_comma("Over\\Under");
assert_eq!(result, "Over\\Under");
```

**Comment by NGA-TRAN** [comment]
@alamb 
> What is the story for the data already in the production database?

This is a good point. In order for us to fix this panic, we need to wipe the catalog again. Or we have to remove/reset all affected information from the catalog (partition keys of the affected partitions) and reload affected data which I think better to wipe the catalog at this stage.

> I don't really understand this choice. I think this PR would contain less code and be less likely to have corner cases if it used the csv crate.

I have thought about this. I agree less code and less likely to have corner cases if it used the csv crate. However, I have the opposite concern that we would introduce new escapes to data that is not commas. I think it is less likely we hit that. 

How about both of us think a bit about this and if on Monday we do not find anything harm from cvs crate in the case of non-comma, then I will use csv crate?

**Comment by alamb** [comment]
> This is a good point. In order for us to fix this panic, we need to wipe the catalog again.

Maybe now is a good time to write a tool that can wipe just a part of a production database (e.g. a namespace or specific partitions) 🤔 

> I have thought about this. I agree less code and less likely to have corner cases if it used the csv crate. However, I have the opposite concern that we would introduce new escapes to data that is not commas. I think it is less likely we hit that.

I guess I figured that since we would be reading/writing with the same library, if non-comma data was escaped that would be ok as it will survive the roundtrip  🤔 

I will think about this more over the weekend for sure

**Comment by domodwyer** [comment]
It might be better to store the array of column names as an [explicit array in postgres](https://www.postgresql.org/docs/current/arrays.html) rather than having to deal with encoding/decoding potentially ambiguous string-based solutions, dealing with escaping of the delimiter, etc. We're effectively trying to turn an array into a string, and turn it back into an array when we read it again - we should just avoid converting it into a string entirely!

I think we could change the `partition` table to be something like the following, where we store the `sort_key` array as a native array data type:

```sql
CREATE TABLE IF NOT EXISTS partition (
    sort_key TEXT[],

    # ...other columns...
);
```

I am happy to do this if it is helpful :+1:

**Comment by alamb** [comment]
> we should just avoid converting it into a string entirely!

If we never need to access the sortkey via SQL in postgres, I think we could go a step farther and just encode directly using protobuf or something else.

I have been assuming that there is value in storing the sort key in postgres in a form that can be easily queried and interpreted via ad hoc human issued queries (aka when we are "poking" with the production catalog). 

If the "should be easy to understand in adhoc manual queries" postgres array sounds like a good solution to me

**Comment by alamb** [comment]
> Having to juggle proto serialisations adds a lot of friction 

100% agree -- it is good to have confirmed this

**Comment by NGA-TRAN** [comment]
Close this without merging. We went with different solution in #4801

**Pull request closed without merging**